### PR TITLE
Support snippets in /usr/lib/security

### LINKED ISSUE #23: [RFE] pwquality.conf.d: please support snippets in /usr/lib/security too
In the context of stateless systems and immutables OS (e.g. CoreOS flavors of Redhat and Fedora, RHCOS/FCOS), it is worth to leave the content of `/etc` under exclusive ownership of the user, and move all system defaults to an immutable and OS-owned location under `/usr`.

To exploit the full potential of that, it would be helpful that libpwquality would start looking up configuration fragments in `/usr/lib/security/pwquality.conf.d/` too, with a lower priority compared to `/etc`.
That way, vendors/distributions can be ship and update their own OS defaults without affecting user customizations.

### PR DESCRIPTION
**Pull request by ListeriaM**
Fix #23 

I'm using `signed char` instead of `bool` because I wasn't sure if I should `#include <stdbool.h>`.

### CODE REVIEW

File: src/settings.c [hunk 1]
```diff
@@ -20,6 +20,9 @@
 #include "pwquality.h"
 #include "pwqprivate.h"
 
+static const char pwquality_base_cfgfile[] = PWQUALITY_BASE_CFGFILE;
+static char pwquality_default_cfgfile[PATH_MAX] = PWQUALITY_DEFAULT_CFGFILE;
+
```
Comment (t8m) [inline src/settings.c:25]: These values should be inside the pwq struct.

Overriding the config file name should happen inside both base and default config paths.

The use case is for having separate configuration for different applications and they should be able to have a separate configuration in /usr/lib.

File: src/pwquality.h [hunk 2]
```diff
@@ -93,6 +93,10 @@ int
 pwquality_read_config(pwquality_settings_t *pwq, const char *cfgfile,
         void **auxerror);
 
+/* Set the default configuration file for use by pwquality_read_config() */
+void
+pwquality_set_default_config_name(const char *cfgfile);
```
Comment (t8m) [inline src/pwquality.h:98]: This will need to have a `pwquality_settings_t *pwq` argument.

File: src/settings.c [hunk 3]
```diff
@@ -251,15 +305,52 @@ pwquality_read_config(pwquality_settings_t *pwq, const char *cfgfile, void **aux
...
                 char *subcfg;
 
                 if (rv) {
                         free(namelist[i]);
                         continue;
                 }
 
-                if (asprintf(&subcfg, "%s/%s", dirname, namelist[i]->d_name) < 0)
+                if (isbase != NULL && isbase[i])
+                        subcfgdir = PWQUALITY_BASE_CFGFILE ".d";
```
Comment (t8m) [inline src/settings.c:352]: Unfortunately this will need changes as well.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by listeriaceae** [comment]
Maybe `PWQ_SETTING_BASE_CFGFILE` and `PWQ_SETTING_DEFAULT_CFGFILE` should be added to `pwquality_get_str_value()` as well? But I don't know if there's any merit to it, or to adding them to `s_map[]` and `pwquality_set_str_value()` and ditching `pwquality_set_config_name()` altogether, or to having different functions for setting `base_cfgfile` and `default_cfgfile`.

It's not easy to make what seem like small decisions in a library that isn't my own, but feel free to edit anything you would have done differently.

**Comment by t8m** [comment]
To me the config file name is a kind of meta-setting. So IMO using a separate API call for it is worth it.

**Comment by listeriaceae** [comment]
Fair enough, then if there's nothing else (like squashing commits or anything) this should be ready to merge.

**Comment by t8m** [comment]
I need to think this trough whether we need to do some fine-tuning of this PR. Unfortunately I am currently busy with other things.

**Comment by t8m** [comment]
I am sorry for not getting into this yet. I am still too busy with other work.

**Comment by listeriaceae** [comment]
That's fine, I didn't mean to rush you, just figured I'd let you know I was done with it.

File: src/settings.c [hunk 4]
```diff
@@ -279,6 +386,47 @@ pwquality_read_config(pwquality_settings_t *pwq, const char *cfgfile, void **aux
...
+        case PWQ_SETTING_DEFAULT_CFGFILE:
+                free(pwq->default_cfgfile);
+                pwq->default_cfgfile = dup;
+                break;
+        default:
+                free(dup);
+                return PWQ_ERROR_UNKNOWN_SETTING;
+        }
+
+        return 0;
```
Comment (t8m) [inline src/settings.c:427]: My idea was something like this:

```suggestion
#define PWQUALITY_BASE_PATH "/usr/lib/security"
#define PWQUALITY_DEFAULT_PATH "/etc/security"
#define PWQUALITY_DEFAULT_NAME "pwquality.conf"
/* change the default configuration file name */
int
pwquality_set_config_name(pwquality_settings_t *pwq,
        const char *cfgname)
{
        char *base_path, *default_path;

        if (cfgname == NULL)
                cfgname = PWQUALITY_DEFAULT_NAME;

        if (asprintf("&base_path, "%s/%s", PWQUALITY_BASE_PATH, cfgname) < 0)
                return PWQ_ERROR_MEM_ALLOC;
        if (asprintf("&default_path, "%s/%s", PWQUALITY_DEFAULT_PATH, cfgname) < 0) {
                free(base_path);
                return PWQ_ERROR_MEM_ALLOC;
        }
        free(pwq->base_cfgfile);
        pwq->base_cfgfile = base_path;
        free(pwq->default_cfgfile);
        pwq->default_cfgfile = default_path;
        return 0;
```

**Review by t8m — CHANGES_REQUESTED** [review]
I think just allowing to change the file name but not the full paths should be sufficient.

File: src/settings.c (hunk 4 above)
Comment (listeriaceae) [reply src/settings.c]: I don't think it's a good idea to enforce the use of the `security` directory for the default config files.

**Comment by ListeriaM** [comment]
Did you notice I ammended the commit when I made my last comment?

If you're on the fence on wether or not to allow users to change the base and default config paths separately we can defer the addition of the `pwquality_set_config_name()` API to a different PR and have this one be just the merging of the default and base configurations in `pwquality_read_config()`. That way it's not a breaking change if you later decide to allow it.

**Comment by t8m** [comment]
@ListeriaM I am really sorry for taking so long time. What about a compromise of changing pwquality_set_config_name() to be as in my proposal above but the default argument would be "security/pwquality.conf" and not just "pwquality.conf"? Not sure if there would be a better name for the function though.

**Review by t8m — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged. Thank you for your contribution and patience.

**Pull request merged**